### PR TITLE
Implement Default for MaybeUninit

### PR DESCRIPTION
**Pull request by petertodd**
Note how if implemented this removes an unsafe in the array initialization example in the `MaybeUninit` main docs, as with `Default` you can just create the array directly. This also could serve as an alternate API concept to the current (unstable) `Foo::<T>::uninit()` stuff, at least for the sized types. Though even if `uninit()` is stabilized I think they still makes sense to implement.

I guess I'd need to open a tracking issue for this?

### CODE REVIEW

File: src/libcore/mem/maybe_uninit.rs [hunk 1]
```diff
@@ -230,6 +230,29 @@ impl<T: Copy> Clone for MaybeUninit<T> {
     }
 }
 
+#[unstable(feature = "maybe_uninit_default", issue = "0")]
```
Comment (Centril) [inline src/libcore/mem/maybe_uninit.rs]: Implementations are insta-stable.

**Comment by Centril** [comment]
cc @RalfJung

**Comment by Centril** [comment]
> Though even if `uninit()` is stabilized I think they still makes sense to implement.

`uninit()` is stable?

**Comment by petertodd** [comment]
> > Though even if `uninit()` is stabilized I think they still makes sense to implement.
> 
> `uninit()` is stable?

Oh, I meant the `Box::uninit()` etc. stuff.

File: src/libcore/mem/maybe_uninit.rs (hunk 1 above)
Comment (petertodd) [reply src/libcore/mem/maybe_uninit.rs]: Good point. Fixed.

**Comment by RalfJung** [comment]
No safety concerns from my side. I guess there is a question here if things like `MaybeUninit::<u32>::default()` could reasonably be expected to be zero-initialized instead of uninitialized? Generally I feel like this can obscure unsafe code using uninitialized memory and make it harder to read (as often happens when relying on trait impls), which seems dangerous.

But I'll happily leave that decision to T-libs.

**Comment by petertodd** [comment]
OTOH most types for which zero is a valid value don't need MaybeUninit: they probably already have cheap default constructors.

**Comment by RalfJung** [comment]
`[u8; N]` for sizable N is a prime use-case for uninitialized memory and can be zero-initialized. I think that is a counter-example to your statement.

**Comment by petertodd** [comment]
Another argument is if we decide that uninit was the wrong default, we can change it to zeroed later and still be backwards compatible; we can't go the other direction.

**Comment by petertodd** [comment]
*Can* be zero initialized. But easier to just initialize with [0;N] or make use of 'u8::default()'

Meanwhile for the stuff where you really need uninit, having to zero initialize means the initialization is no longer a no-op, which can be much more more expensive. 

The final 'assume_init()' is unsafe anyway, and in practice this stuff comes up more often in generic code that you write once. (at least, that's been my experience)

**Comment by RalfJung** [comment]
> easier to just initialize with [0;N] or make use of 'u8::default()'

What I am saying is that this type regularly "needs" `MaybeUninit` for performance.

> in practice this stuff comes up more often in generic code that you write once. (at least, that's been my experience)

Seems like an argument to optimize for readability then, which `default` generally does not.

**Comment by petertodd** [comment]
On October 10, 2019 1:00:31 AM GMT+03:00, Ralf Jung <notifications@github.com> wrote:
>> easier to just initialize with [0;N] or make use of 'u8::default()'
>
>What I am saying is that this type regularly "needs" `MaybeUninit` for
>performance.

Not sure what you mean there. Can you give an example of a code pattern?

>> in practice this stuff comes up more often in generic code that you
>write once. (at least, that's been my experience)
>
>Seems like an argument to optimize for readability then, which
>`default` generally does not.

So I ran into this with code manipulating generic containers, and default was just a nice way to make the c

**Comment by petertodd** [comment]
On October 10, 2019 1:00:31 AM GMT+03:00, Ralf Jung <notifications@github.com> wrote:
>> easier to just initialize with [0;N] or make use of 'u8::default()'
>
>What I am saying is that this type regularly "needs" `MaybeUninit` for
>performance.
>
>> in practice this stuff comes up more often in generic code that you
>write once. (at least, that's been my experience)
>
>Seems like an argument to optimize for readability then, which
>`default` generally does not.

(Whoops, hit send too early!)

So I ran into this with code manipulating generic containers in-place, and default was a nice way to make the container not care about the fact that it was being initialized as uninit, then init in-place. Made for a bit less unsafe code.

In that case defaulting to uninit made sense as it's a true zero cost allocation (modulo the memory itself of course).

Note that I'm considering submitting another pull-req / RFC for an assume_ininit() trait once I get more experience myself on what API makes sense (I have a project that basically has arena allocation as a sub goal). In the meantime I thought that implementing Default deemed like an obvious improvement that makes sense regardless of how you build API's around it.

**Comment by RalfJung** [comment]
> Not sure what you mean there. Can you give an example of a code pattern?

The usual one is something like `let buffer: [u8; 4096] = mem::uninitialized();` followed by `something.read(&mut buffer);` (from the `Read` trait).

> Made for a bit less unsafe code.

Wait, how can adding this `Default` instance reduce the amount of unsafe code? Initializing a `MaybeUninit` is safe.

**Comment by petertodd** [comment]
@RalfJung So the example given in the pull-req is one: when initializing an array with `MaybeUninit`, `Default` means can omit one unsafe to create the array, leaving just the actual unsafety of asserting that it's initialized.

**Comment by RalfJung** [comment]
> when initializing an array with MaybeUninit, Default means can omit one unsafe to create the array

No unsafe needed:
```rust
let mut uninit = [MaybeUninit::<bool>::uninit(); 10];
```

**Comment by petertodd** [comment]
> > when initializing an array with MaybeUninit, Default means can omit one unsafe to create the array
> 
> No unsafe needed:
> 
> ```rust
> let mut uninit = [MaybeUninit::<bool>::uninit(); 10];
> ```

That actually depends on a `Copy` bound; wouldn't work for a different type (which does say maybe I should have a different example!).

**Comment by Centril** [comment]
> That actually depends on a `Copy` bound; wouldn't work for a different type (which does say maybe I should have a different example!).

https://github.com/rust-lang/rust/issues/49147

**Comment by rkruppe** [comment]
> #49147

`MaybeUninit::uninit()` isn't rvalue promotable, though.

That is not to say I endorse this impl, I'm skeptical for the same reasons as @RalfJung. Perhaps the array-of-uninit use case can be made safe with other, more contained, API additions?

**Comment by Centril** [comment]
> `MaybeUninit::uninit()` isn't rvalue promotable, though.

(Stick it in a `const`; that's what we do internally in libcore, https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#associatedconstant.UNINIT)

**Comment by rkruppe** [comment]
Oh, fair point. But besides not being very discoverable, you can't even create the const if type parameters are involved.

**Comment by RalfJung** [comment]
> MaybeUninit::uninit() isn't rvalue promotable, though.

We *might* be less conservative with promotion for array initializers as unlike normal promotion, here we do not have the scenario that if we promote too aggressively, we could end up breaking code.

> That actually depends on a Copy bound; wouldn't work for a different type (which does say maybe I should have a different example!).

Ah, and `Default` for arrays calls the lower `Default` N times?
That seems like a hack to me, exploiting the "lifting to arrays"-nature of the `Default` impl when really we'd like something more targeted.

**Comment by petertodd** [comment]
> > That actually depends on a `Copy` bound; wouldn't work for a different type (which does say maybe I should have a different example!).
> 
> #49147

I'm looking forward to that getting implemented. But it's been over a year. :)

Somewhat unrelated: what does `clone()`/copy do with uninitialized bytes anyway? I was a bit surprised to find `MaybeUninit` implemented `Clone`/`Copy`. I'd hope that LLVM would keep track of which bytes within a `MaybeUninit` were initialized and only skip copying them if they truly were undef. But I don't see any tests for this (grep 'MaybeUninit' -r -C 10000 src/test  | grep -i clone doesn't turn up anything).

**Comment by Centril** [comment]
> I'm looking forward to that getting implemented. But it's been over a year. :)

(It is implemented... sorta? Or do you mean stabilized?)

**Comment by RalfJung** [comment]
> Somewhat unrelated: what does clone()/copy do with uninitialized bytes anyway?

Same as `memcpy`. The new memory is just as uninitialized as the old one.

(Remember that memory in the sense we are talking about here has, for every address, an `Option<u8>` to also indicate whether this byte is initialized or not. *That* is the data that is being copied.)

> I'd hope that LLVM would keep track of which bytes within a MaybeUninit were initialized and only skip copying them if they truly were undef. But I don't see any tests for this

There is no such test as `MaybeUninit` is a normal `union` without any magic. We also don't have a codegen test specifically for `Vec` to see if LLVM doesn't screw up `memcpy` there...

Yes we do rely on LLVM not being stupid when doing `memcpy`, including for partially initialized data. But so does *all of the rest if Rust* and most of C/C++ as well (e.g. when copying/assigning a struct with padding, which typically will not be initialized), so... I think we are good.

**Comment by petertodd** [comment]
> > I'm looking forward to that getting implemented. But it's been over a year. :)
> 
> (It is implemented... sorta? Or do you mean stabilized?)

Oh, I just saw the issue, which still says it's not implemented.

**Comment by Centril** [comment]
> Oh, I just saw the issue, which still says it's not implemented.

Fixed. :)

**Comment by petertodd** [comment]
> > Somewhat unrelated: what does clone()/copy do with uninitialized bytes anyway?
> 
> Same as `memcpy`. The new memory is just as uninitialized as the old one.

I mean, Rust doesn't have `memcpy`, at least as a public API (I know internally it does).

See I would, for example, assume that the uninitialized bytes are not guaranteed to be the *same* in the source and destination, as being undefined the compiler is free to skip actually copying them. That's not actually documented; might be worth adding a note about that to the `Clone` impl for people less paranoid than me. :)

> (Remember that memory in the sense we are talking about here has, for every address, an `Option<u8>` to also indicate whether this byte is initialized or not. _That_ is the data that is being copied.)
> 
> > I'd hope that LLVM would keep track of which bytes within a MaybeUninit were initialized and only skip copying them if they truly were undef. But I don't see any tests for this
> 
> There is no such test as `MaybeUninit` is a normal `union` without any magic. We also don't have a codegen test specifically for `Vec` to see if LLVM doesn't screw up `memcpy` there...
> 
> Yes we do rely on LLVM not being stupid when doing `memcpy`, including for partially initialized data. But so does _all of the rest if Rust_ and most of C/C++ as well (e.g. when copying/assigning a struct with padding, which typically will not be initialized), so... I think we are good.

Probably the best response to my question would be we *can't* test that because we're not making any testable promises: we only guarantee that the initialized bytes will be copied over; uninitialized bytes could be any value.

That said, I have written serialization-type code that, for example, assumes that coercing a `&mut MaybeUninit<T>` to a raw byte pointer (`*mut [u8]`) and writing to those bytes causes the bytes for the `T` type to be considered initialized (being careful to make sure all used bytes are actually written). That strikes me as the type of thing that might lead to problems in a way that, say, `Vec` doesn't run into, as the types are different.

**Comment by petertodd** [comment]
> > That actually depends on a Copy bound; wouldn't work for a different type (which does say maybe I should have a different example!).
> 
> Ah, and `Default` for arrays calls the lower `Default` N times?

Yup.

Right now it's an exception to the arbitrary sized const generics impls that were recently merged too, so the code is actually a bunch of `Default` calls "in parallel". But that'd be fairly easy to fix at the cost of a bit of complexity around handling drop properly on panic (no rush given the 32 element limit).

> That seems like a hack to me, exploiting the "lifting to arrays"-nature of the `Default` impl when really we'd like something more targeted.

Huh, I gotta admit to my taste that seems fairly natural: you're initializing an array of `MaybeUninit` just like you would any other type. Equally you can do `Vec::resize_with(MaybeUninit::default)` and a whole bunch of similar patterns.

**Comment by memoryruins** [comment]
> I mean, Rust doesn't have memcpy, at least as a public API

https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html

**Comment by petertodd** [comment]
> > I mean, Rust doesn't have memcpy, at least as a public API
> 
> https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html

It's not documented how that interacts with uninitialized bytes.

**Comment by RalfJung** [comment]
> It's not documented how that interacts with uninitialized bytes. I'm aware it says it's "semantically equivalent" to memcpy. But when I read that I'm not assuming that equivalence extended to behavior around uninitialized data.

Not sure why you wouldn't. But then, strictly speaking I have no idea what the behavior of `memcpy` for uninit data is in C (the standard isn't helping), so to be fair the comparison isn't really explaining anything.

But, the semantics "Copies count * size_of::<T>() bytes from src to dst" *does* actually explain everything. It's just that without a proper concrete memory model ("a byte is an `Option<u8>` tracking initializedness") it's hard to say anything more concrete.

(I should also add the disclaimer that really, a byte is *not* just `Option<u8>`. It's [more complicated than that](https://www.ralfj.de/blog/2018/07/24/pointers-and-bytes.html).)

> See I would, for example, assume that the uninitialized bytes are not guaranteed to be the same in the source and destination, as being undefined the compiler is free to skip actually copying them.

They are the same. Both are `None`. But I don't think that's what you meant. ;)

But your question shows you are thinking at the wrong level. You [cannot explain the behavior of uninitialized memory by thinking of random data](https://www.ralfj.de/blog/2019/07/14/uninit.html). You have to think at the level of the Rust Abstract Machine. And and that level, the src and target bytes *are* the same. Including uninitialized bytes.

---

So improve the `copy` and `copy_nonoverlapping` docs, I propose we add a sentence like:

> In particular, if any byte of the `src` range was uninitialized, the corresponding byte in the `dest` range is uninitialized after the copy.

**Comment by JohnTitor** [comment]
Ping from triage: @petertodd any updates on this?

**Comment by petertodd** [comment]
@JohnTitor I'm going to close this. I think there's a lot of good points made by others above on better ways to do what I anticipated this being used for; the "uninit" API's being added elsewhere are probably a better overall approach. Also given that `MaybeUninit` is used with unsafe code making it so easy to derive default constructors has a risk of violating assumptions in structs using `MaybeUninit` - easy to miss a "default" in a derive.

@RalfJung Thanks for the detailed reply!

**Pull request closed without merging**